### PR TITLE
Remove previously deprecated `--enable-test-discovery`

### PR DESCRIPTION
**Pull request by MaxDesiatov**
This flag was deprecated for 4 years since Swift 5.4, Swift 6.0 is a good opportunity to remove it.

Closes https://github.com/apple/swift-package-manager/issues/7389.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Sources/CoreCommands/Options.swift
Comment (grynspan) [inline Sources/CoreCommands/Options.swift:467]: We may want to keep this flag in place and just not use it for anything except generating a diagnostic. At least for a little bit longer, to give affected packages like swift-numerics a chance to migrate.

File: Sources/CoreCommands/SwiftCommandState.swift
Comment (grynspan) [inline Sources/CoreCommands/SwiftCommandState.swift:375]: Ditto.

File: Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift
Comment (grynspan) [inline Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift:77]: A comment here explaining why this remains in place might be helpful for future maintainers.

Comment (MaxDesiatov) [reply Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift:77]: I think this will be gone in a future commit in this same PR, we don't expect build parameters serialization to be compatible between versions.

Comment (grynspan) [reply Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift:77]: Okay! I'm fine with that.

**Review by stmontgomery — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@yim-lee @ktoso I see a few repositories you worked on still have `LinuxMain.swift`. Especially if any of those had `fatalError` in them requiring users to pass `--enable-test-discovery`, it's best to delete those files at this point and use plain `swift test` when Swift 6.0 is released.

**Comment by ktoso** [comment]
Sounds good, we only support swift 5.7+ nowadays in most telos so we can drop it I think.

Swift log I think supports ancient versions but we can maybe stop testing on them or something... I'll think about it

**Comment by yim-lee** [comment]
Thanks for the heads-up @MaxDesiatov.

> I see a few repositories you worked on still have `LinuxMain.swift`. 

Is `LinuxMain.swift` ignored in newer Swift versions?

> Especially if any of those had `fatalError` in them requiring users to pass `--enable-test-discovery`

We have repos that pass `--enable-test-discovery` even in newer versions, but I don't think any of them does `fatalError`. Nevertheless, we will adjust the build params as needed once this change gets merged.

@ktoso I guess we should consider bumping up the Swift version requirement if we run into trouble removing `--enable-test-discovery`/`LinuxMain.swift`. At some point it doesn't make much sense to support ancient Swift versions.

**Review by yim-lee — APPROVED** [review]

**Comment by yim-lee** [comment]
Got it. Thanks @MaxDesiatov.

**Comment by MaxDesiatov** [comment]
> Don't we want to _stop_ using it?

IDK, if we do we'd have to go through a deprecation cycle with those for at least a few releases first.

**Pull request merged**
